### PR TITLE
Image Management

### PR DESCRIPTION
**Pull request by ficast**
ACs

- User creates/updates a GraphGist Candidate with links out to images
-  Images are automatically imported (into a service like imagekit.io)
-  Links to the images are automatically replaced by links to imagekit.io (or equivalent)
-  Image nodes are created and linked to the User
-  User can see "my images" and delete images

Example use: 
- User creates a graphgist by copy+pasting some content
- User realizes an image they didn't want to share has now been uploaded to imagekit. 
- User deletes the link from the graphgist but it's still there!
- User goes to "my gallery" and deletes the image node
- User checks the link, which is now broken (as expected)

Obs:
- Sometimes when user goes to "my gallery" and deletes the image it can still be showed for a while, probably caused by some delay with imagekit service (DNS, etc). I've tested and sometimes it is instantly removed and sometimes not.

### CODE REVIEW

File: api/src/graphgists/mutations.js [hunk 1]
```diff
@@ -69,7 +98,23 @@ export const CreateGraphGist = async (root, args, context, info) => {
     );
     const authorPerson = authorResult.records[0].get('p').properties;
 
-    var categoryUuid;
+    //imagekit
+    for (let imagekitImage of imagekitImages) {
+      await txc.run(
+        `
+        MATCH (g:GraphGist {uuid: $uuid})<-[:IS_VERSION]-(gc:GraphGistCandidate), (a:Person {uuid: $authorUuid})
+        CREATE (i:Image $image)
```
Comment (whatSocks) [inline api/src/graphgists/mutations.js:106]: 🤔

**Comment by Masudbdqa** [comment]
Looks good.

**Review by Masudbdqa — APPROVED** [review]
Looks good

**Comment by ficast** [comment]
https://www.loom.com/share/22b9633114824200a453dc8dbaecae65

File: ui-react/src/App.js [hunk 2]
```diff
@@ -74,23 +76,23 @@ const GET_TOOLBAR = gql`
       name
       num_graphgists
       image(first: 1) {
-        source_url
+        source
```
Comment (nossila) [inline ui-react/src/App.js:79]:
```suggestion
        source_url
```

revert change

File: ui-react/src/App.js [hunk 3]
```diff
@@ -74,23 +76,23 @@ const GET_TOOLBAR = gql`
...
       }
     }
     industries: Industry {
       slug
       name
       num_graphgists
       image(first: 1) {
-        source_url
+        source
```
Comment (nossila) [inline ui-react/src/App.js:87]:
```suggestion
        source_url
```

revert change

File: ui-react/src/App.js [hunk 4]
```diff
@@ -74,23 +76,23 @@ const GET_TOOLBAR = gql`
...
       }
     }
     challenges: Challenge {
       slug
       name
       num_graphgists
       image(first: 1) {
-        source_url
+        source
```
Comment (nossila) [inline ui-react/src/App.js:95]:
```suggestion
        source_url
```

revert change

File: ui-react/src/candidates/CandidateCard.js [hunk 5]
```diff
@@ -105,23 +105,24 @@ GraphGistCandidateCard.fragments = {
...
       }
       categories(first: 3) {
         __typename
         uuid
         slug
         name
-        image(first: 1) {
-          source_url
+        image {
+          source
```
Comment (nossila) [inline ui-react/src/candidates/CandidateCard.js:118]:
```suggestion
        source_url
```

revert change

File: ui-react/src/candidates/CandidateCard.js [hunk 6]
```diff
@@ -105,23 +105,24 @@ GraphGistCandidateCard.fragments = {
...
         }
       }
       author {
         name
         slug
         image(first: 1) {
-          source_url
+          source
```
Comment (nossila) [inline ui-react/src/candidates/CandidateCard.js:125]:
```suggestion
        source_url
```

revert change

File: api/src/images/types.js [hunk 7]
```diff
@@ -6,12 +6,15 @@ export const Image = {
...
+      url = null
     } = obj.source_url.properties;
+    if (source === 'imagekit') return url;
     const size = "medium";
     const partitions = uuid
       .match(/.{9}/g)[0]
       .match(/.{1,3}/g)
       .join("/");
-    return `https://${AWS_S3_BUCKET_NAME}.s3.amazonaws.com/graph_starter/images/sources/${partitions}/${size}/${source_file_name}?${source_updated_at}`;
+      return `https://${AWS_S3_BUCKET_NAME}.s3.amazonaws.com/graph_starter/images/sources/${partitions}/${size}/${source_file_name}?${source_updated_at}`;
```
Comment (nossila) [inline api/src/images/types.js:18]:
```suggestion
    return `https://${AWS_S3_BUCKET_NAME}.s3.amazonaws.com/graph_starter/images/sources/${partitions}/${size}/${source_file_name}?${source_updated_at}`;
```

identation

**Review by nossila — CHANGES_REQUESTED** [review]

**Review by nossila — APPROVED** [review]

**Comment by whatSocks** [comment]
@ficast can you also update the readme with how to use the imagekit stuff so newcomers to the project can know how to work around it or use it

**Comment by whatSocks** [comment]
Screen Shot 2021-07-01 at 10 46 26 AM

little edge case - if I upload a .mov in the image portion of the admin edit, i get this broken image situation

Screen Shot 2021-07-01 at 10 47 06 AM

for a solution could we allow only image formats in this field?

**Review by whatSocks — APPROVED** [review]

**Pull request merged**